textattack/attack_recipes/clare_li_2020.py

```diff
@@ -0,0 +1,108 @@
+)
+
+from .attack_recipe import AttackRecipe
+
+
+class CLARE2020(AttackRecipe):
+    """Li, Zhang, Peng, Chen, Brockett, Sun, Dolan.
+
+    "Contextualized Perturbation for Textual Adversarial Attack" (Li et al., 2020)
+
+    https://arxiv.org/abs/2009.07502
+
+    This method uses greedy search with replace, merge, and insertion transformations that leverage a
+    pretrained language model. It also uses USE similarity constraint.
+    """
+
+    @staticmethod
+    def build(model):
+        # "This paper presents CLARE, a ContextuaLized AdversaRial Example generation model
+        # that produces fluent and grammatical outputs through a mask-then-infill procedure.
+        # CLARE builds on a pre-trained masked language model and modifies the inputs in a context-aware manner.
+        # We propose three contex-tualized  perturbations, Replace, Insert and Merge, allowing for generating outputs of
+        # varied lengths."
+        #
+        # "We  experiment  with  a  distilled  version  of RoBERTa (RoBERTa_{distill}; Sanh et al., 2019)
+        # as the masked language model for contextualized infilling."
+        # Because BAE and CLARE both use similar replacement papers, we use BAE's replacement method here.
+
+        shared_masked_lm = transformers.AutoModelForCausalLM.from_pretrained(
+            "distilroberta-base"
+        )
+        shared_tokenizer = transformers.AutoTokenizer.from_pretrained(
+            "distilroberta-base"
+        )
+        transformation = CompositeTransformation(
+            [
+                WordSwapMaskedLM(
+                    method="bae",
+                    masked_language_model=shared_masked_lm,
+                    tokenizer=shared_tokenizer,
+                    max_candidates=20,
+                    min_confidence=5e-4,
+                ),
+                WordInsertionMaskedLM(
+                    masked_language_model=shared_masked_lm,
+                    tokenizer=shared_tokenizer,
+                    max_candidates=20,
+                    min_confidence=5e-4,
+                ),
+                WordMergeMaskedLM(
+                    masked_language_model=shared_masked_lm,
+                    tokenizer=shared_tokenizer,
+                    max_candidates=20,
+                    min_confidence=5e-4,
+                ),
+            ]
+        )
+
+        #
+        # Don't modify the same word twice or stopwords.
+        #
+        constraints = [RepeatModification(), StopwordModification()]
+
+        # "A  common  choice  of sim(·,·) is to encode sentences using neural networks,
+        # and calculate their cosine similarity in the embedding space (Jin et al., 2020)."
+        # The original implementation uses similarity of 0.7. Since the CLARE code is based on the TextFooler code,
+        # we need to adjust the threshold to account for the missing / pi in the cosine similarity comparison
+        #  So the final threshold is 1 - (1 - 0.7) / pi =  0.904507034.
```

Comment:
So my original threshold doesn't have the PI in cosine similarity. Not sure whether it has some significant impacts.

Reply:
I think the issue is whether you used cosine similarity or angular similarity. Did you use cosine similarity? Then it should just be 0.7.

Reply:
Yes, I just used cosine similarity